indigo_libs/indigo_client.c

```diff
@@ -444,10 +444,14 @@ static void *server_thread(indigo_server_entry *server) {
 }
 
 indigo_result indigo_connect_server(const char *name, const char *host, int port, indigo_server_entry **server) {
+	return indigo_connect_server_with_interface(name, host, port, 0, server);
+}
+
+indigo_result indigo_connect_server_with_interface(const char *name, const char *host, int port, uint32_t interfaceIndex, indigo_server_entry **server) {
 	int empty_slot = used_server_slots;
 	pthread_mutex_lock(&mutex);
 	for (int dc = 0; dc < used_server_slots; dc++) {
-		if (indigo_available_servers[dc].thread_started && !strcmp(indigo_available_servers[dc].host, host) && indigo_available_servers[dc].port == port) {
+		if (indigo_available_servers[dc].thread_started && !strcmp(indigo_available_servers[dc].host, host) && indigo_available_servers[dc].port == port && indigo_available_servers[dc].interfaceIndex == interfaceIndex) {
```

Comment:
The interface_index member is used here, when you identify a server (or better a connection?) in the available_servers list. In the new scenario we need a connection per interface.

Reply:
I get it... I think :) maybe in that case you can use some other name as this is not related to any particular interface, maybe some other name would be better - connection_index or connection_id. And add some more useful explanation when and how to use it.

As I understand it, it is used to permit multiple connections to a single service and be able to distinguish them and it has nothing to do with the physical interface. Is that correct?

Do not get me wrong I just want to make sure everything is clear.

Reply:
> I get it... I think :) maybe in that case you can use some other name as this is not related to any particular interface, maybe some other name would be better - 

Actually it is, the DNS Service Discovery API returns the POSIX interface index which uniquely identifies a physical network device (e.g.index in  https://www.man7.org/linux/man-pages/man3/if_nametoindex.3.html). Therefore I think it makes sense to keep this name, since it is not my invention. But I agree, maybe the name indigo_available_servers is not appropriate anymore. With my change the elements of the array have now the meaning of a connection ...


> As I understand it, it is used to permit multiple connections to a single service and be able to distinguish them 

Yes

>and it has nothing to do with the physical interface. Is that correct?

As mentioned above, the interface index is related to a physical index. So e.g. LAN networks device is mapped to index 1 and WLAN network deice is mapped to index 2, index 0 is never used. So my change enables a client application process to hold several connections (each network device has its own IP address) to the same indigo server. On the server side for each network device a separate worker thread is started, that all works out-of-the box. The change on indigo side was minimal. Very good architecure, congrats! 

> Do not get me wrong I just want to make sure everything is clear.

I absolutely appreciate this discussion, maybe I have missed some important point...

Reply:
Ok I get the point, but this number can be virtually everything and not limited to interface index. So in your app it may be device index, but for some reason my app may want to start several connections and each must have a separate number in this field (the framework does not care what is in there as long as they are different) so I would prefer to rename it to something more general like connection_id or connection_index. So you can store device index in your case, I may store sequential number, a random number or something else... it is up to the client to handle it correctly. The framework will just refuse to connect with the same same number to the same service twise. So from the framework point of view it is id or index, but not interface. And my argument is that some more general name would be better. @polakovic  please share your opinion!

Reply:
I understand and fully agree. I'll send a new commit this evening. Thanks Rumen!

Reply:
Hi, I'd like to involve a little bit, but I'm still at work, pls. let ma day or two... :(

Reply:
I have another concern, this will break the binary compatibility with indigo shared library, as we add a new field in the structure. All existing clients need to be recompiled otherwise they will crash which may be an issue. We need to discuss this with Peter and evaluate the risks... Sorry, we are very reluctant about changing the core... 
However i like your idea and we will find a way accommodate your change somehow...

Reply:
@rumengb : You are right, I went into exactly that issue when I did an upgrade of the indigo libs last time. Nevertheless I pushed your renaming suggestion so that the  discussion and the code is in a consistent state.

@polakovic : OK, there is no hurry, take your time.